group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -2427,25 +2471,259 @@ private boolean hasMemberSubscriptionChanged(
+            return bumpGroupEpoch;
+        }
+
+        // Adjust the count to account for the updated member.
+        Map<String, Integer> subscribedRegularExpressions = new HashMap<>(group.subscribedRegularExpressions());
+        if (oldSubscribedTopicRegex != null) {
+            subscribedRegularExpressions.compute(oldSubscribedTopicRegex, Utils::decValue);
+        }
+        if (newSubscribedTopicRegex != null) {
+            subscribedRegularExpressions.compute(newSubscribedTopicRegex, Utils::incValue);
+        }
+
+        // If the number of subscribed regexes is different from the number of resolved ones,
+        // we must trigger a refresh.
+        requireRefresh |= subscribedRegularExpressions.size() != group.numResolvedRegularExpressions();
+
+        // If the last time that regexes were refreshed is older than the refresh interval,
+        // we must trigger a refresh.
+        requireRefresh |= time.milliseconds() >
+            group.lastResolvedRegularExpressionRefreshTimeMs() + consumerGroupRegularExpressionRefreshInternalMs;
+
+        if (requireRefresh) {
+            Set<String> regexes = subscribedRegularExpressions.keySet();
+            executor.schedule(
+                key,
+                () -> refreshRegularExpressions(groupId, log, time, metadataImage, regexes),
+                (result, exception) -> handleRegularExpressionsResult(groupId, result, exception)
+            );
+        }
+
+        return bumpGroupEpoch;
+    }
+
+    /**
+     * Resolves the provided regular expressions. Note that this static method is executed
+     * as an asynchronous task in the executor. Hence, it should not access any state from
+     * the manager.
+     *
+     * @param groupId   The group id.
+     * @param log       The log instance.
+     * @param time      The time instance.
+     * @param image     The metadata image to use for listing the topics.
+     * @param regexes   The list of regular expressions that must be resolved.
+     * @return The list of resolved regular expressions.
+     */
+    private static Map<String, ResolvedRegularExpression> refreshRegularExpressions(
+        String groupId,
+        Logger log,
+        Time time,
+        MetadataImage image,
+        Set<String> regexes
+    ) {
+        log.debug("[GroupId {}] Refreshing regular expressions: {}", groupId, regexes);
+        Map<String, Set<String>> resolvedRegexes = new HashMap<>(regexes.size());
+        List<Pattern> compiledRegexes = new ArrayList<>(regexes.size());
+        for (String regex : regexes) {
+            resolvedRegexes.put(regex, new HashSet<>());
+            try {
+                compiledRegexes.add(Pattern.compile(regex));
+            } catch (PatternSyntaxException ex) {
+                // This should not happen because the regular expressions are validated
+                // when received from the members. If for some reasons, it would
+                // happen, we log it and ignore it.
```

Comment:
if this should not happen, should we throw an illegal state exception?

also nit: "If for some reason,.."

Reply:
I chose to log and ignore for two reasons: 1) i don't want to block resolving other regexes; 2) throwing an exception would just log it somewhere else (in CoordinatorExecutorImpl).